File: matlab/+hohmann_het/Propulsion.m
% Author: A Taylor | Purpose: Hall Effect Thruster propulsion models | Ref: Vallado/Curtis
classdef Propulsion
% PROPULSION - High-fidelity Hall Effect Thruster (HET) models.
%
% Models anode efficiency, beam voltage, exhaust velocity, specific impulse,
% mass flow rate, and propellant budget.
%
% Reference:
%   Goebel & Katz, "Fundamentals of Electric Propulsion," JPL/Wiley, 2008.
%   Vallado, "Fundamentals of Astrodynamics and Applications," 4th ed.
%
% Usage:
%   state = hohmann_het.Propulsion.compute_het_state(300, 1350, 0.50)
%   mp    = hohmann_het.Propulsion.propellant_mass(1000, 3.8526, 1514)
%
% Author: A Taylor | Ref: Vallado/Curtis

    properties (Constant, Access = private)
        %% Physical constants
        G0       = 9.80665;          % Standard gravity [m/s^2]
        E_CHARGE = 1.602176634e-19;  % Elementary charge [C]
        M_XE     = 2.180174e-25;     % Xenon atom mass [kg]  (131.293 u)
    end

    methods (Static)

        function state = compute_het_state(discharge_voltage, discharge_power, anode_efficiency)
        % COMPUTE_HET_STATE  Compute HET operating point from discharge parameters.
        %
        % Equations (LaTeX):
        %
        %   v_e = sqrt(2 * eta_a * e * V_d / m_Xe)   [beam exhaust velocity, m/s]
        %
        %   I_sp = v_e / g_0                           [specific impulse, s]
        %
        %   F = 2 * eta_a * P_d / v_e                 [thrust, N]
        %
        %   m_dot = F / v_e                            [mass flow rate, kg/s]
        %
        % Inputs:
        %   discharge_voltage  - anode discharge voltage [V],  positive scalar
        %   discharge_power    - input discharge power   [W],  positive scalar
        %   anode_efficiency   - anode efficiency eta_a in (0,1), scalar
        %
        % Output:
        %   state - struct with fields:
        %       isp              [s]
        %       exhaust_velocity [m/s]
        %       thrust           [N]
        %       mass_flow        [kg/s]
        %       anode_efficiency [-]
        %       discharge_power  [W]

            arguments
                discharge_voltage  (1,1) double {mustBePositive, mustBeFinite}
                discharge_power    (1,1) double {mustBePositive, mustBeFinite}
                anode_efficiency   (1,1) double {mustBeInRange(anode_efficiency, 0, 1, "exclusive", "exclusive")}
            end

            G0       = hohmann_het.Propulsion.G0;
            E_CHARGE = hohmann_het.Propulsion.E_CHARGE;
            M_XE     = hohmann_het.Propulsion.M_XE;

            ve   = sqrt(2.0 * anode_efficiency * E_CHARGE * discharge_voltage / M_XE);
            isp  = ve / G0;
            F    = 2.0 * anode_efficiency * discharge_power / ve;
            mdot = F / ve;

            state.isp              = isp;
            state.exhaust_velocity = ve;
            state.thrust           = F;
            state.mass_flow        = mdot;
            state.anode_efficiency = anode_efficiency;
            state.discharge_power  = discharge_power;
        end

        function mp = propellant_mass(m_initial, delta_v_km_s, isp)
        % PROPELLANT_MASS  Tsiolkovsky rocket equation.
        %
        %   m_p = m_0 * (1 - exp(-Delta_v / (g_0 * I_sp)))
        %
        % Inputs:
        %   m_initial    - initial spacecraft mass [kg],  positive scalar
        %   delta_v_km_s - required delta-V [km/s],       non-negative scalar
        %   isp          - specific impulse [s],           positive scalar
        %
        % Output:
        %   mp - propellant mass [kg]

            arguments
                m_initial    (1,1) double {mustBePositive,    mustBeFinite}
                delta_v_km_s (1,1) double {mustBeNonnegative, mustBeFinite}
                isp          (1,1) double {mustBePositive,    mustBeFinite}
            end

            G0 = hohmann_het.Propulsion.G0;
            ve = isp * G0;                      % m/s
            dv = delta_v_km_s * 1000.0;         % km/s -> m/s
            mp = m_initial * (1.0 - exp(-dv / ve));
        end

        function tb = burn_time(thrust, mass_flow, delta_v_km_s, m_initial)
        % BURN_TIME  Constant-thrust burn time for a given delta-V.
        %
        %   T_b = (m_0 / m_dot) * (1 - exp(-Delta_v / v_e))
        %
        % Inputs:
        %   thrust       - thruster force [N],      positive scalar
        %   mass_flow    - mass flow rate [kg/s],   positive scalar
        %   delta_v_km_s - required delta-V [km/s], non-negative scalar
        %   m_initial    - initial mass [kg],        positive scalar
        %
        % Output:
        %   tb - burn time [s]

            arguments
                thrust       (1,1) double {mustBePositive,    mustBeFinite}
                mass_flow    (1,1) double {mustBePositive,    mustBeFinite}
                delta_v_km_s (1,1) double {mustBeNonnegative, mustBeFinite}
                m_initial    (1,1) double {mustBePositive,    mustBeFinite}
            end

            ve = thrust / mass_flow;                % m/s
            dv = delta_v_km_s * 1000.0;             % km/s -> m/s
            tb = (m_initial / mass_flow) * (1.0 - exp(-dv / ve));
        end

    end % methods (Static)
end % classdef

File: python/src/hohmann_het/propulsion.py
# Author: A Taylor | Purpose: Hall Effect Thruster propulsion models | Ref: Vallado/Curtis
"""High-fidelity Hall Effect Thruster (HET) propulsion module.

Models anode efficiency, beam voltage, exhaust velocity, specific impulse,
mass flow rate, and propellant budget.

References
----------
- Goebel & Katz, "Fundamentals of Electric Propulsion," JPL/Wiley, 2008.
- Vallado, "Fundamentals of Astrodynamics and Applications," 4th ed.
"""
from __future__ import annotations

import math
from dataclasses import dataclass

import astropy.units as u

# ---------------------------------------------------------------------------
# Physical constants
# ---------------------------------------------------------------------------
#: Standard gravity [m/s^2]
G0: u.Quantity = 9.80665 * u.m / u.s**2

#: Elementary charge [C]
E_CHARGE: float = 1.602176634e-19

#: Xenon atom mass [kg]  (131.293 u, 1 u = 1.66053906660e-27 kg)
M_XE: float = 2.180174e-25


# ---------------------------------------------------------------------------
# Result dataclass
# ---------------------------------------------------------------------------
@dataclass(frozen=True)
class HETState:
    """Hall Effect Thruster operating-point state.

    Attributes
    ----------
    isp : u.Quantity
        Specific impulse [s].
    exhaust_velocity : u.Quantity
        Effective exhaust velocity [m/s].
    thrust : u.Quantity
        Thrust force [N].
    mass_flow : u.Quantity
        Propellant mass flow rate [kg/s].
    anode_efficiency : float
        Anode efficiency (dimensionless).
    discharge_power : u.Quantity
        Discharge power [W].
    """

    isp: u.Quantity
    exhaust_velocity: u.Quantity
    thrust: u.Quantity
    mass_flow: u.Quantity
    anode_efficiency: float
    discharge_power: u.Quantity


# ---------------------------------------------------------------------------
# Public functions
# ---------------------------------------------------------------------------
def compute_het_state(
    discharge_voltage: "u.Quantity | float",
    discharge_power: "u.Quantity | float",
    anode_efficiency: float,
) -> HETState:
    r"""Compute Hall Effect Thruster operating point from discharge parameters.

    **Beam-limited exhaust velocity** (Goebel & Katz, Eq. 7-13):

    .. math::
        v_e = \sqrt{\frac{2 \eta_a e V_d}{m_{\rm Xe}}}

    **Specific impulse:**

    .. math::
        I_{sp} = \frac{v_e}{g_0}

    **Thrust** (power balance):

    .. math::
        F = \frac{2 \eta_a P_d}{v_e}

    **Mass flow rate:**

    .. math::
        \dot{m} = \frac{F}{v_e}

    Parameters
    ----------
    discharge_voltage : astropy.units.Quantity or float
        Anode discharge voltage [V]. If plain ``float``, assumed V.
    discharge_power : astropy.units.Quantity or float
        Input discharge power [W]. If plain ``float``, assumed W.
    anode_efficiency : float
        Anode efficiency :math:`\eta_a \in (0, 1)`.

    Returns
    -------
    HETState
        Complete HET operating point with units.

    Raises
    ------
    ValueError
        If voltage, power, or efficiency are out of range.
    """
    if not isinstance(discharge_voltage, u.Quantity):
        discharge_voltage = float(discharge_voltage) * u.V
    if not isinstance(discharge_power, u.Quantity):
        discharge_power = float(discharge_power) * u.W

    V_d = discharge_voltage.to(u.V).value
    P_d = discharge_power.to(u.W).value

    if V_d <= 0.0 or P_d <= 0.0:
        raise ValueError("Discharge voltage and power must be positive.")
    if not (0.0 < anode_efficiency < 1.0):
        raise ValueError("Anode efficiency must be in (0, 1).")

    ve   = math.sqrt(2.0 * anode_efficiency * E_CHARGE * V_d / M_XE)  # m/s
    isp  = ve / G0.value                                                # s
    F    = 2.0 * anode_efficiency * P_d / ve                           # N
    mdot = F / ve                                                       # kg/s

    return HETState(
        isp=isp * u.s,
        exhaust_velocity=ve * (u.m / u.s),
        thrust=F * u.N,
        mass_flow=mdot * (u.kg / u.s),
        anode_efficiency=anode_efficiency,
        discharge_power=P_d * u.W,
    )


def propellant_mass(
    m_initial: "u.Quantity | float",
    delta_v: "u.Quantity | float",
    isp: "u.Quantity | float",
) -> u.Quantity:
    r"""Propellant mass from the Tsiolkovsky rocket equation.

    .. math::
        m_p = m_0 \left(1 - \exp\!\left(-\frac{\Delta v}{g_0 \, I_{sp}}\right)\right)

    Parameters
    ----------
    m_initial : astropy.units.Quantity or float
        Initial (wet) spacecraft mass [kg]. If plain ``float``, assumed kg.
    delta_v : astropy.units.Quantity or float
        Required delta-V. If plain ``float``, assumed km/s.
    isp : astropy.units.Quantity or float
        Specific impulse [s]. If plain ``float``, assumed s.

    Returns
    -------
    u.Quantity
        Propellant mass [kg].
    """
    if not isinstance(m_initial, u.Quantity):
        m_initial = float(m_initial) * u.kg
    if not isinstance(delta_v, u.Quantity):
        delta_v = float(delta_v) * (u.km / u.s)
    if not isinstance(isp, u.Quantity):
        isp = float(isp) * u.s

    m0   = m_initial.to(u.kg).value
    dv   = delta_v.to(u.m / u.s).value   # convert km/s -> m/s
    isp_s = isp.to(u.s).value

    ve = isp_s * G0.value                 # m/s
    mp = m0 * (1.0 - math.exp(-dv / ve))
    return mp * u.kg


def burn_time(
    thrust: "u.Quantity | float",
    mass_flow: "u.Quantity | float",
    delta_v: "u.Quantity | float",
    m_initial: "u.Quantity | float",
) -> u.Quantity:
    r"""Constant-thrust burn time for a given delta-V.

    Derived from integrating the rocket equation at constant mass flow:

    .. math::
        T_b = \frac{m_0}{\dot{m}}
              \left(1 - \exp\!\left(-\frac{\Delta v}{v_e}\right)\right)

    where :math:`v_e = F / \dot{m}`.

    Parameters
    ----------
    thrust : astropy.units.Quantity or float
        Thruster force [N]. If plain ``float``, assumed N.
    mass_flow : astropy.units.Quantity or float
        Mass flow rate [kg/s]. If plain ``float``, assumed kg/s.
    delta_v : astropy.units.Quantity or float
        Required delta-V. If plain ``float``, assumed km/s.
    m_initial : astropy.units.Quantity or float
        Initial spacecraft mass [kg]. If plain ``float``, assumed kg.

    Returns
    -------
    u.Quantity
        Burn time [s].
    """
    if not isinstance(thrust, u.Quantity):
        thrust = float(thrust) * u.N
    if not isinstance(mass_flow, u.Quantity):
        mass_flow = float(mass_flow) * (u.kg / u.s)
    if not isinstance(delta_v, u.Quantity):
        delta_v = float(delta_v) * (u.km / u.s)
    if not isinstance(m_initial, u.Quantity):
        m_initial = float(m_initial) * u.kg

    F    = thrust.to(u.N).value
    mdot = mass_flow.to(u.kg / u.s).value
    dv   = delta_v.to(u.m / u.s).value
    m0   = m_initial.to(u.kg).value

    if F <= 0.0 or mdot <= 0.0 or m0 <= 0.0 or dv < 0.0:
        raise ValueError("Thrust, mass flow, and initial mass must be positive; dv >= 0.")

    ve = F / mdot                                    # m/s
    tb = (m0 / mdot) * (1.0 - math.exp(-dv / ve))
    return tb * u.s
